src/metabase/actions/error.clj

```diff
@@ -0,0 +1,17 @@
+(ns metabase.actions.error)
+
+(def violate-unique-constraint
+  "Error type for SQL unique constraint violation."
+  ::violate-unique-constraint)
+
+(def violate-not-null-constraint
+  "Error type for SQL not null constraint violation."
+  ::violate-not-null-constraint)
+
+(def violate-foreign-key-constraint
+  "Error type for SQL foreign key constraint violation."
+  ::violate-foreign-key-constraint)
+
+(def incorrect-value-type
+  "Error type for SQL incorrect value type."
+  ::incorrect-value-type)
```

Comment:
My motivation for defining var for these things are:
- we can provide docs
- easy to find all error types
- centralized and maybe in the future we can do sth like error hierarchy as we have in `metabase.query-processor.error-type`

I could remove this ns if you find it's not justified. normal keywords also work for me.

Reply:
It might be useful to add another layer of indirection and centralize the creation of the error maps currently returned by each implementation of `maybe-parse-sql-error`. For each error-type, we need to return the same format of data under the `:message` and `:errors` keys. We should only need to define the `:message` and error messages under `:errors` in one place for each error-type, not once for each combination of driver and error-type. Then the only thing you would need to define in each driver's maybe-parse-sql-error implementation is how to extract the required components from the DBs error message.

Reply:
@calherries 
Initially, I went with that approach and I really hope we could.

but the problem is the error message from each driver does not provide the same level of granularity.

For example for incorrect type error.
Mysql error message: `Incorrect integer value: 'STRING' for column "table.id" at row 1`
Postgres error message: `invalid input syntax for type integer: \"S\"`

Notice the error message from Postgres does not contain which column it has type error. 
So if we want the message and errors to be the same for all drivers, it enforces the error message to be generic and requires as few components as possible.

That's why I changed to "error" per driver approach like this, it's extra effort, but it gives more flexibility.